distributed/cluster.py

```diff
@@ -2,65 +2,252 @@
+    output_queue = Queue()
+    cmd_dict = {'cmd': cmd, 'label': label, 'address': worker_addr, 'port': worker_port,
+                'input_queue': input_queue, 'output_queue': output_queue}
+
+    # Start the thread
+    thread = Thread(target=async_ssh, args=[cmd_dict])
+    thread.start()
+
+    return merge(cmd_dict, {'thread': thread})
 
-    return {'client': ssh, 'channel': channel}
 
 
 class Cluster(object):
-    def __init__(self, center, workers):
-        self.center = assoc(start_center(center), 'address', center)
-        self.workers = [assoc(start_worker(center, worker), 'address', worker)
-                        for worker in workers]
-        sleep(1)
-        self.report()
-
-    def report(self):
-        self.center['channel'].settimeout(1)
-        print("Center\n------")
-        print(self.center['channel'].recv(10000).decode())
-        for worker in self.workers:
-            channel = worker['channel']
-            address = worker['address']
-            channel.settimeout(1)
-            print("Worker: %s\n---------" % address+ '-' * len(address))
-            print(channel.recv(10000).decode())
+    def __init__(self, center_addr, center_port, worker_addrs, workers_per_node = 1, cpus_per_worker = None, logdir = None):
+
+        self.center_addr = center_addr
+        self.center_port = center_port
+        self.workers_per_node = workers_per_node
+        self.cpus_per_worker = cpus_per_worker
+
+        # Generate a universal timestamp to use for log files
+        import datetime
+        if logdir is not None:
+            logdir = os.path.join(logdir, "dcluster_" + datetime.datetime.now().strftime("%Y-%m-%d_%H:%M:%S"))
+            print(bcolors.WARNING + 'Output will be redirected to logfiles stored locally on individual woker nodes under "{logdir}".'.format(logdir=logdir) + bcolors.ENDC)
+        self.logdir = logdir
+
+        # Keep track of all running threads
+        self.threads = []
+
+        # Start the center node
+        self.center = start_center(logdir, center_addr, center_port)
+
+        # Start worker nodes
+        self.workers = []
+
+        # For rate limiting (below)
+        import multiprocessing
+        ncores = multiprocessing.cpu_count()
+
+        for i, addr in enumerate(worker_addrs):
+
+            # Rate limit the starting of worker nodes. We don't want to fire off
+            # too many SSH connections at the same time, or this machine will
+            # briefly grind to a halt.
+            sleep(float( i // (2*ncores) ))
+
+            self.add_worker(addr)
+
+    def monitor_remote_processes(self):
+
+        # Form a list containing all processes, since we treat them equally from here on out.
+        all_processes = [self.center] + self.workers
+
+        try:
+            while(1):
+                for process in all_processes:
+                    while not process['output_queue'].empty():
+                        print(process['output_queue'].get())
```

Comment:
I was thinking about this last night.  It might be better to have a single output_queue that all threads write to.  They're threadsafe and so you can just do 

```python
while true:
    print(queue.get())
```